sysutils/SilentKnight/Portfile

```diff
@@ -0,0 +1,40 @@
+# -*- coding: utf-8; mode: tcl; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- vim:fenc=utf-8:ft=tcl:et:sw=4:ts=4:sts=4
+
+PortSystem          1.0
+
+name                SilentKnight
+version             1.6
+categories          sysutils aqua
+platforms           darwin
+supported_archs     x86_64
+license             freeware
+maintainers         {@lhaeger gmx.net:lothar.haeger} openmaintainer
+
+description         Automatic checking of security systems
+long_description    SilentKnight checks your Mac's basic security systems are active, reports version numbers \
+                    of security configuration files which are active, the latest updates installed, and makes it easy \
+                    to check for and install updates. Ideal for checking that SIP is enabled, and it has Apple's latest \
+                    silent security updates. Also checks EFI firmware, security settings and data files, and has both \
+                    a summary Help page and a detailed reference.
+
+homepage            https://eclecticlight.co/lockrattler-systhist/
+
+master_sites        https://eclecticlightdotcom.files.wordpress.com/2020/01/
+distname            [strsed [string tolower ${name}${version}] {g/[.]//}]
+use_zip             yes
+
+checksums           rmd160  215b5897bf7b6f82ed00747c87b39482a17b73ed \
+                    sha256  b0584a3adacc0911305e15953ec7a018df08db82606d765fc7acced3d03577a9 \
+                    size    8907933
+
+use_configure       no
+
+build {}
```

Comment:
Given this and `supported_archs  x86_64`, is this port simply downloading a binary? Can it be changed to build it instead?

Reply:
It's freeware but not open source, so no, it can't be built from source, unfortunately.

Reply:
I would prefer that, too, but it's not always an option. So "we" can either:

- reject the port for not being OSS
- accept it for being useful free software
- what else?

Are there some published requirements (not preferences, I know them by now) for new ports to be accepted somewhere? 
Since this is coming up again and again and I do not have the time to create ports that will not get accepted anyway, I'd rather have a set of rules than run the same discussion over and over again.

Re: "redistribute" - all that MacPorts distributes is the Portfile, the software itself is downloaded from the author's website. I think there was a way to make MacPorts never distribute a port in binary form, which could be added to such ports, so they do not "cost" any bandwidth, hosting fees etc. 

Other than that I do not see any reason why a useful piece of software should not be searchable, installable and updateable via MacPorts, even if it can't be built from source. Such ports merely automate the above steps, which is doubtlessly helpful to anyone using MacPorts already and may even be a reason for some to try it out. 
In the end most people do not care much about how MacPorts makes it happen, they just want a hassle-free tool to manage software that's not available on AppStore.

Reply:
> Re: "redistribute" - all that MacPorts distributes is a 1.77 KB Portfile, the software itself is downloaded from the author's website.

That's not true: MacPorts also mirrors and redistributes the original distfile and, if the license indicates the binary is distributable, it hosts and distributes the binary.

> I think there was a way to make MacPorts never distribute a port in binary form, which could be added to such ports, so they do not "cost" any bandwidth, hosting fees etc. Will see if I can find it...

Specifying a license that is not known to MacPorts to be distributable will prevent us from distributing the binary. Using `NoMirror` in the license field will additionally prevent us from distributing the distfile, but you should not use it unless the developer specifically states that they do not allow us to do so.

Reply:
> MacPorts also mirrors and redistributes the original distfile and, if the license indicates the binary is distributable, it hosts and distributes the binary.

Hmm, I might be doing something wrong then, but it seems I always end up compiling from source, maybe I'm too fast with installing updates (both macOS/Xcode and ports) to be able to find pre-compiled ports (I do not use "port -s" unless preparing a new port or updating one, and macports.conf neither has it set as a default).

> Specifying a license that is not known to MacPorts to be distributable will prevent us from distributing the binary. Using `NoMirror` in the license field will additionally prevent us from distributing the distfile, but you should not use it unless the developer specifically states that they do not allow us to do so.

But wouldn't that be a good way to handle ports that come as binary only and basically help user to search/install/update rather than patch/configure/compile? A portgroup could set this together with build{} and use_configure=no, any maybe also handle destroot {...} in a prettier way.

I'm still wondering about the reluctance of accepting ports like this on the ground of not being OSS. I understand the roots of MacPorts, but we're 20+ years onwards since it all starts and there's so much usefull but hard to find (and keep up to date) stuff out there, where MacPorts could help without much cost/effort and shine for it's capabilites and flexibility.

Reply:
> it seems I always end up compiling from source

There are probably a dozen possible reasons why that might be the case. If you'd like to diagnose it, maybe write to macports-users.

> > Specifying a license that is not known to MacPorts to be distributable will prevent us from distributing the binary. Using `NoMirror` in the license field will additionally prevent us from distributing the distfile, but you should not use it unless the developer specifically states that they do not allow us to do so.
> 
> But wouldn't that be a good way to handle ports that come as binary only and basically help user to search/install/update rather than patch/configure/compile? A portgroup could set this together with build{} and use_configure=no, any maybe also handle destroot {...} in a prettier way.

Huh? I don't see any need for a portgroup here, and if you're suggesting that it is a good idea for ports that just install a binary to specify a restrictive and/or a no-mirror license, then I don't understand what you think that will help.

> I'm still wondering about the reluctance of accepting ports like this on the ground of not being OSS. I understand the roots of MacPorts, but we're 20+ years onwards since it all starts and there's so much usefull but hard to find (and keep up to date) stuff out there, where MacPorts could help without much cost/effort and shine for it's capabilites and flexibility.

I am not opposed to adding this port.

Primary strengths of MacPorts are the ability to build projects from source and manage dependencies. If a thing is already built, and if it is not a dependency of another thing, then there is much less necessity of having that thing in MacPorts. But another strength of MacPorts is that a user can come to one place to install a lot of different software, and on that strength it still makes sense to allow software even if it is binary-only.

Reply:
> Huh? I don't see any need for a portgroup here.

I do not see a "need" either, but it would help shorten the Portfile and provide useful defaults for this use case. Automating and reducing redundancy is generally a good thing in my experience.

> another strength of MacPorts is that a user can come to one place to install a lot of different software, and on that strength it still makes sense to allow software even if it is binary-only.

That's exactly my line of thinking and why I added this port. Maybe I've been working too long in a field where we the end user experience is the ultimate proof of the pudding.

> if you're suggesting that it is a good idea for ports that just install a binary to specify a restrictive and/or a no-mirror license, then I don't understand what you think that will help.

If the reluctance to accept ports like this was (at least partly) based on not wanting to waste MacPorts' resources on it, limiting the "cost" to distributing the Portfile only might help general acceptance. 
I understand that MacPorts cannot re-distribute all the useful non-AppStore software out there that would deserve a wider audience, there's simply too much of it. So if a port falls into the category "nice and useful, but binary-only and not the kind we would usually include without reluctance", this could be a solution to reduce it's impact on the MacPorts environment while still making it available for users to search/install/update.
A no-mirror license would have that effect, I understand. Or did I get that wrong again?

Reply:
> I do not see a "need" either, but it would help shorten the Portfile and provide useful defaults for this use case. Automating and reducing redundancy is generally a good thing in my experience.

Obviously. But how valuable would it really be to have a portgroup that just contains the two lines:

```
use_configure no
build {}
```

You can then shorten the portfile from those two lines to a single portgroup line. Is that really so much better? Are there really so many ports that need that?

If you are thinking that the portgroup would do more than that, what more would it do? If you are thinking it would install the files in the destroot phase, how do you propose that the portgroup should know what files to install and where?


> the end user experience is the ultimate proof of the pudding

Obviously yes we want to do things that make life easier for our users.


> If the reluctance to accept ports like this was (at least partly) based on not wanting to waste MacPorts' resources on it, limiting the "cost" to distributing the Portfile only might help general acceptance.
> I understand that MacPorts cannot re-distribute all the useful non-AppStore software out there that would deserve a wider audience, there's simply too much of it. So if a port falls into the category "nice and useful, but binary-only and not the kind we would usually include without reluctance", this could be a solution to reduce it's impact on the MacPorts environment while still making it available for users to search/install/update.

I don't have any disk space or bandwidth concerns about our distfiles and packages. We have in the past lost one of our mirrors because they could not accommodate our disk usage, but if that is really a concern we have other ways of addressing it, like more regularly running our cleanup scripts that delete old distfiles and packages that are no longer being used. We should do that, rather than arbitrarily marking some ports for non-mirroring and non-distribution of binaries.


> A no-mirror license would have that effect, I understand. Or did I get that wrong again?

As I said, NoMirror in the license field will cause us not to mirror the distfile (which in the case of this port means the binary as originally packaged by its developer). And a non-redistributable license in the license field will cause us not to distribute our binary (which in the case of this port means our repackaged binary).

Reply:
> I don't have any disk space or bandwidth concerns about our distfiles and packages. We have in the past lost one of our mirrors because they could not accommodate our disk usage, but if that is really a concern we have other ways of addressing it, like more regularly running our cleanup scripts that delete old distfiles and packages that are no longer being used. We should do that, rather than arbitrarily marking some ports for non-mirroring and non-distribution of binaries.

Ok, then my assumptions were wrong, great to hear. 

I'd still think that resource usage could be optimized for binary-only ports, be that supported by a portgroup or otherwise. Take a look at dbeaver-community: a ~50MB binary download, cached in http://nue.de.distfiles.macports.org/dbeaver-community/ 12 versions in cache > ~600MB 
Then 10 of those versions can be found repackaged for 9 OS versions each in http://nue.de.packages.macports.org/dbeaver-community/  - that's 4.5 GB of pure redundancy. And those packages would not even speed up installation on clients significantly or reduce download size.

A portgroup could help prevent those 4.5 GB by adding a non-distributable license automatically (or flagging the port in some other way so the build scripts ignore it).

Reply:
> If you are thinking it would install the files in the destroot phase, how do you propose that the portgroup should know what files to install and where?

I'd thought about providing a default copy scheme in `destroot{}` , probably slightly different depending on archive type and port category. 
DMGs usually have the binary stored in their root folder, other archives usually expand to a subfolder named like the archive and the binary is found there. App bundles (category contains "aqua") could default to `${name}.app` and go into `${destroot}${applications_dir}` while command line stuff ends up in `${destroot}/bin` by default.
If the above does not apply, overriding `destroot{}` would be required.